shared/protocol/src/main/java/io/pravega/shared/protocol/netty/CommandEncoder.java

```diff
@@ -89,64 +83,43 @@ protected void encode(ChannelHandlerContext ctx, Object msg, ByteBuf out) throws
+                currentBlockSize = Math.max(TYPE_PLUS_LENGTH_SIZE, blockSizeSupplier.getAppendBlockSize());
+                bytesLeftInBlock = currentBlockSize;
+                segmentBeingAppendedTo = append.segment;
+                writeMessage(new AppendBlock(session.id), out);
+                if (ctx != null) {
+                    ctx.executor().schedule(new Flusher(ctx.channel(), currentBlockSize),
+                                            blockSizeSupplier.getBatchTimeout(),
+                                            TimeUnit.MILLISECONDS);
                 }
+            }
 
-                session.lastEventNumber = append.getEventNumber();
-                session.eventCount++;
-                ByteBuf data = append.getData();
-                int msgSize = TYPE_PLUS_LENGTH_SIZE + data.readableBytes();
-                // Is there enough space for a subsequent message after this one?
-                if (bytesLeftInBlock - msgSize > TYPE_PLUS_LENGTH_SIZE) {
-                    bytesLeftInBlock -= serializeEvent(new Event(data), new ByteBufOutputStream(out));
-                } else {
-                    ByteArrayOutputStream bout = new ByteArrayOutputStream();
-                    serializeEvent(new Event(data), bout);
-                    byte[] serializedMessage = bout.toByteArray(); 
-                    int bytesInBlock = bytesLeftInBlock - TYPE_PLUS_LENGTH_SIZE;
-                    ByteBuf dataInsideBlock = wrappedBuffer(serializedMessage, 0, bytesInBlock);
-                    ByteBuf dataRemainging = wrappedBuffer(serializedMessage,
-                                                           bytesInBlock,
-                                                           serializedMessage.length - bytesInBlock);
-                    writeMessage(new PartialEvent(dataInsideBlock), out);
-                    writeMessage(new AppendBlockEnd(session.id,
-                                                    currentBlockSize - bytesLeftInBlock,
-                                                    dataRemainging,
-                                                    session.eventCount,
-                                                    session.lastEventNumber,
-                                                    0L), out);
-                    bytesLeftInBlock = 0;
-                    session.eventCount = 0;
-                }
+            session.lastEventNumber = append.getEventNumber();
+            session.eventCount++;
+            ByteBuf data = append.getData();
+            int msgSize = data.readableBytes();
+            // Is there enough space for a subsequent message after this one?
+            if (bytesLeftInBlock - msgSize > TYPE_PLUS_LENGTH_SIZE) {
+                out.writeBytes(data);
+                bytesLeftInBlock -= msgSize;
+            } else {
+                int bytesInBlock = bytesLeftInBlock - TYPE_PLUS_LENGTH_SIZE; 
+                ByteBuf dataInsideBlock = data.readSlice(bytesInBlock);
```

Comment:
We should also invoke `data.retain()` here right?

Reply:
Its not needed. dataInsideBlock is just a temporary variable that goes away after line 114. So we could call retain here and release 3 lines down. But it would not change anything because `data` is still being held the whole time.